core/src/mender-api.c

```diff
@@ -98,11 +103,15 @@ mender_err_t
 mender_api_init(mender_api_config_t *config) {
 
     assert(NULL != config);
-    assert(NULL != config->artifact_name);
     assert(NULL != config->device_type);
     assert(NULL != config->host);
     mender_err_t ret;
 
+    /* Load and set artifact_name here */
+    if ((MENDER_OK != mender_storage_get_artifact_name(&artifact_name)) && (NULL != artifact_name)) {
```

Comment:
This is incorrect. If `provides` contains artifact_name, that one should take precedence. That is why in the acceptance criteria of the task I suggested to have a function that reads the two store keys and merge the results, so that the artifact_name is only one and is consistent after load.

Does it make sense?

Reply:
But the artifact_name from provides does take precedence now? This function loads it from the store, and the artifact name stored is either from the deployment, or the one from provides if provides-depends is defined. That is why I overwrite the `artifact_name` key in the deployment data; it is overwritten by the artifact_name from provides. Perhaps it should be a separate key like `provides_artifact_name` ? I didn't see a reason to merge artifact_name from provides into new_provides only to delete it again before storing, if that makes sense.

Reply:
Wait, I may have written that comment too fast. It looks like the logic changed a bit after I addressed some comments and made some changes, I'll need to look at it again to see how it works :sweat_smile:

Reply:
Discussed f2f and found an agreement. Waiting for the rework :smiley: